### ISSUE TITLE
Slow code execution for some features

### ISSUE THREAD

**Issue by ymonb1291** [denoland/deno#8543]
I found a couple of methods that run significantly slower in Deno compared to Node.

Below is a little program that runs both in both runtimes. It tests a number of functions 1 million times each and reports the time it took to execute each of them. I'm testing basic methods (such as Array.map, for loop, Date.now(), ...), but I'm also interested in the performance modules of both frameworks.
* For Deno, I'm talking about the [performance user timing APIs](https://github.com/denoland/deno/pull/6421)
* For Node, it's the [perf_hooks](https://nodejs.org/api/perf_hooks.html) module

```
const list = [1, 2, 3, 4, 5, 6, 7, 8, 9];
const obj1 = {foo: 1};
const obj2 = {bar: 1};
const json = JSON.stringify(obj1);

// Declare the performance tool (different between Node and Deno)
let perf;
try {
  // Use perf_hooks in Node
  perf = require("perf_hooks").performance;
} catch {
  // Use performance in Deno
  perf = performance
}

// Declare the test functions
function mapFn() {return list.map(num => num*10);}
function reduceFn() {return list.reduce((a, b) => a+b);}
function forEachFn() {return list.forEach(num => num*10);}
function sliceFn() {return list.slice(0, 5);}
function forFn() {for (let i = 0; i < list.length; i++) {list[i]*10};}
function objectAssignFn() {return Object.assign({}, obj1, obj2)};
function jsonStringifyFn() {return JSON.stringify(obj1)};
function jsonParseFn() {return JSON.parse(json)};
function performanceMarkFn() {perf.mark()};
function performanceNowFn() {perf.now()};
function dateNowFn() {Date.now()};

// Loop through each test function and measure the time it takes
// to execute 1 million iterations.
[
  mapFn,
  reduceFn,
  forEachFn,
  sliceFn,
  forFn,
  dateNowFn,
  objectAssignFn,
  jsonStringifyFn,
  jsonParseFn,
  performanceNowFn,
  performanceMarkFn,
]
  .forEach(fn => {
    let start = Date.now();
    for (let i = 0; i < 1000000; i++) {
      fn();
    }
    let end = Date.now();
    console.log(`${fn.name} - Time elapsed: ${end-start} ms`);
  });
```

Running this code gives the following results in Node v15.2.0:
```
mapFn - Time elapsed: 80 ms
reduceFn - Time elapsed: 26 ms
forEachFn - Time elapsed: 27 ms
sliceFn - Time elapsed: 106 ms
forFn - Time elapsed: 36 ms
dateNowFn - Time elapsed: 111 ms
objectAssignFn - Time elapsed: 123 ms
jsonStringifyFn - Time elapsed: 306 ms
jsonParseFn - Time elapsed: 425 ms
performanceNowFn - Time elapsed: 96 ms
performanceMarkFn - Time elapsed: 2548 ms
```
And in Deno 1.5.4:
```
mapFn - Time elapsed: 211 ms
reduceFn - Time elapsed: 32 ms
forEachFn - Time elapsed: 24 ms
sliceFn - Time elapsed: 70 ms
forFn - Time elapsed: 27 ms
dateNowFn - Time elapsed: 133 ms
objectAssignFn - Time elapsed: 102 ms
jsonStringifyFn - Time elapsed: 314 ms
jsonParseFn - Time elapsed: 463 ms
performanceNowFn - Time elapsed: 11288 ms
performanceMarkFn - Time elapsed: 12127 ms
```

We can see that for most of my tests, Deno and Node are quite comparable. However, there are a few major discrepancies:
* **Array.map()** is approx 3x slower on Deno. This result surprises me the most, because other methods such as Array.forEach() and Array.reduce() don't seem to be affected.
* **performance.now()** is approx 110x slower on Deno and **performance.mark()** is approx 5x slower. I understand that the performance module is there to benchmark our code, therefore it shouldn't be that slow.

**Comment by Soremwar** [comment]
@ymonb1291 Are you running the benchmarks with the `--allow-hrtime` flag in Deno? That flag is required in order to obtain accurate times (disabled by default to prevent timing attacks)

**Comment by ymonb1291** [comment]
@Soremwar Yes I've tried that. In the example in the OP I'm measuring duration with `Date.now()`. It's less precise but it doesn't explain these huge differences that I've mentioned.

I can refactor the code to use high resolution timers (see below), but I find that the results are roughly the same despite the use of the `--allow-hrtime` flag.
```
[
  // list of functions
].forEach(fn => {
    let start = perf.now()
    for (let i = 0; i < 1000000; i++) {
      fn();
    }
    let end = perf.now()
    console.log(`${fn.name} - Time elapsed: ${end-start} ms`);
  });
```

**Comment by kevinkassimo** [comment]
For `Array.map()`: this is unlikely our problem. You should probably try logging the v8 version used by Node and Deno. Likely an issue from v8 itself.

For `performance*`: these 2 features, comparing to other functions you tested, involves ops, which are mostly JSON based operations that pass requests and responses between JS/TS and Rust's privileged side, since requesting high resolution timestamp is a privileged operation not doable solely inside of a v8 sandbox. A lot of performance cost concentrate around serialization/deserialization and memory management. Some other fun observation: since there involves permission checks for possible hrtime, this part takes an extra toll on the perf (3%) especially as it processes error message construction. Below is the flamegraph for the test case where we only run `performanceNowFn` 1000000 times. (See svg version [here](https://gist.github.com/kevinkassimo/d1514db1fef01b9d73b178ccd11541be))
![Screen Shot 2020-11-29 at 3 13 19 PM](https://user-images.githubusercontent.com/15007517/100556305-4e781180-3256-11eb-8ebb-0c9ee33d4827.png)

**Comment by lucacasonato** [comment]
We should consider making `performance.now` a minimal op, instead of a JSON op. Looking at this flame graph, we should be able to reduce overhead significantly.

**Comment by bartlomieju** [comment]
@ymonb1291 performance problem with `performance.now()` and `performance.mark()` should be fixed in `1.6.0`. We'll look into using V8 fast API in the future. Tentatively closing.
